File: python/Arrays_Hashing/bucket_sort.py
# implement bucket sort from scratch ( not optimal for arrays with large input values like 1000,1000000 etc)

sample = [5,1,1,4,3,1,0,0,2,4]
def bucket_sort(sample : list) -> list:
    min_val = min(sample)
    max_val = max(sample)
    size = max_val - min_val +1
    bucket = [0] * (size)

    for i in sample:
        bucket[i-min_val] += 1

    k = 0
    for i in range(len(bucket)):
        while bucket[i] > 0:
            sample[k] = i + min_val
            k = k +1
            bucket[i] -= 1

    return sample

    
print(bucket_sort(sample))

File: java/Arrays_Hashing/bucketSort.java
// implement bucket sort from scratch ( not optimal for arrays with large input values like 1000,1000000 etc)


import java.util.*;

public class bucketSort {
    public static void main(String[] args){
        int[] nums = {5,1,1,4,3,1,0,0,2,4};
       
        int max = nums[0];
        int min = nums[0];
        for(int m : nums){
            if(m > max) max = m;
            if(m < min) min = m;
        }
        int size = max - min +1;
        int[] bucket = new int[size];
        int index= 0;
        for(int i = 0 ;i<nums.length ; i++){
            index = nums[i] - min;
            bucket[index]++;
        }

        int k = 0;
        
        for(int i = 0; i<bucket.length;i++){
            while(bucket[i] > 0){
                nums[k] = i + min;
                k++;
                bucket[i]--;
            }
        }
        System.out.println(Arrays.toString(nums));
    }
}